python/mp/rosary.py

```diff
@@ -0,0 +1,158 @@
+import copy
+import threading
+import time
+import math
+
+from mp import color
+from .effects import *
+from pythonosc import udp_client
+from pythonosc import osc_bundle_builder
+from pythonosc import osc_message_builder
+
+class Bead:
+    def __init__(self, index=0):
+        self.index = index
+        self.color = color.Color()
+        
+    def __repr__(self):
+        return "Bead(index={}, color={})".format(self.index, self.color)
+
+    def copy_color(self, color):
+        self.color = copy.copy(color)
+
+
+class Rosary:
+    def __init__(self, ip="127.0.0.1", port=5005):
+        self.beads = []
+        self.bgcolor = color.Color(0,0,0)
+        self.effects = []
+        self.osc_ip = ip
+        self.osc_port = port
+        self.effect_id = 0;
+        self.BEAD_COUNT=60
+        self.run_mainloop = False
+        self.mainloop_delay = 0.03
+        self.effect_registry = {}
+
+        self.osc_client = udp_client.UDPClient(self.osc_ip, self.osc_port)
+
+        for i in range(self.BEAD_COUNT):
+            self.beads.append(Bead(i))
+
+        # some useful predefined sets of beads
+        
+        self.Set_None = frozenset()
+        self.Set_All = frozenset(self.beads)
+        self.Set_Stem = frozenset(self.beads[0:4])
+        self.Set_Ring = frozenset(self.beads[4:60])
+        self.Set_Eighth0 = frozenset(self.beads[4:11])
+        self.Set_Eighth1 = frozenset(self.beads[11:18])
+        self.Set_Eighth2 = frozenset(self.beads[18:25])
+        self.Set_Eighth3 = frozenset(self.beads[25:32])
+        self.Set_Eighth4 = frozenset(self.beads[32:39])
+        self.Set_Eighth5 = frozenset(self.beads[39:46])
+        self.Set_Eighth6 = frozenset(self.beads[46:53])
+        self.Set_Eighth7 = frozenset(self.beads[53:60])
+        self.Set_Quadrent0 = frozenset(self.beads[4:18])
+        self.Set_Quadrent1 = frozenset(self.beads[18:32])
+        self.Set_Quadrent2 = frozenset(self.beads[32:46])
+        self.Set_Quadrent3 = frozenset(self.beads[46:60])
+        self.Set_Half01 = frozenset(self.Set_Quadrent0.union(self.Set_Quadrent1))
+        self.Set_Half12 = frozenset(self.Set_Quadrent1.union(self.Set_Quadrent2))
+        self.Set_Half23 = frozenset(self.Set_Quadrent2.union(self.Set_Quadrent3))
+        self.Set_Half30 = frozenset(self.Set_Quadrent3.union(self.Set_Quadrent0))
+        self.Set_Even_All = frozenset(self.beads[0:60:2])
+        self.Set_Even_Ring = frozenset(self.beads[4:60:2])
+        self.Set_Odd_All = frozenset(self.beads[1:60:2])
+        self.Set_Odd_Ring = frozenset(self.beads[5:60:2])
+
+        # some useful predefined colors
+        self.Color_White = color.Color(1,1,1)
+        self.Color_Red = color.Color(1,0,0)
+        self.Color_Yellow = color.Color(1,1,0)
+        self.Color_Green = color.Color(0,1,0)
+        self.Color_Blue = color.Color(0,0,1)
+        self.Color_Violet = color.Color(1,0,1)
+        self.Color_Cyan = color.Color(0,1,1)
+        self.Color_Black = color.Color(0,0,0)
```

Comment:
Once again, trifling opinion, but both these looks kinda clunky and could use being converted to a dictionary-like thing in the vein of `self.effects`. There's no practical benefit as of now, but this is the over-obsessive side of me prematurely optimizing for some future scenario where we want to return the list of available bead sets or whatever.